### PR TITLE
FileManager: Implement contentsEqual(atPath:andPath:)

### PR DESCRIPTION
**Pull request by spevans**
- Requires some more tests, especially for comparing directories.

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci please test

File: Foundation/FileManager.swift [hunk 1]
```diff
@@ -637,11 +637,161 @@ open class FileManager : NSObject {
...
+        while let item = enumerator1.nextObject() as? String {
+            path1entries.insert(item)
+        }
+
+        while let item = enumerator2.nextObject() as? String {
+            if path1entries.remove(item) == nil {
+                return false
+            }
+        }
+        return path1entries.isEmpty
```
Comment (kperryua) [inline Foundation/FileManager.swift]: I'd expect this to recurse for each item in the two directories. If the contents of 'A/C' doesn't match 'B/C', then this is supposed to fail.

File: Foundation/FileManager.swift [hunk 2]
```diff
@@ -637,11 +637,161 @@ open class FileManager : NSObject {
...
+            guard memcmp(buffer1, buffer2, bytesToRead) == 0 else {
+                return false
+            }
+            bytesLeft -= Int64(bytesToRead)
+        }
+        return true
+    }
+
+    private func _compareSymlinks(withFileSystemRepresentation file1Rep: UnsafePointer<Int8>, andFileSystemRepresentation file2Rep: UnsafePointer<Int8>, size: Int64) -> Bool {
+        let bufSize = Int(size) + 1
```
Comment (kperryua) [inline Foundation/FileManager.swift]: On Darwin at least, `readlink` does not append a NUL byte, so the +1 isn't strictly necessary.

File: Foundation/FileManager.swift [hunk 3]
```diff
@@ -637,11 +637,161 @@ open class FileManager : NSObject {
...
+            return false
+        }
+        let file2Type = file2.st_mode & S_IFMT
+
+        // Are paths the same type: file, directory, symbolic link etc.
+        guard file1Type == file2Type else {
+            return false
+        }
+
+        if file1Type == S_IFCHR {
```
Comment (kperryua) [inline Foundation/FileManager.swift]: The same could be done here for `S_IFBLK`, no?

File: TestFoundation/TestFileManager.swift [hunk 4]
```diff
@@ -620,4 +621,83 @@ class TestFileManager : XCTestCase {
...
+        XCTAssertFalse(fm.contentsEqual(atPath: emptyFile.path, andPath: testDir1.appendingPathComponent("null1").path))
+        XCTAssertFalse(fm.contentsEqual(atPath: emptyFile.path, andPath: testDir1.appendingPathComponent("unreadable_file").path))
+
+        XCTAssertTrue(fm.contentsEqual(atPath: testFile1URL.path, andPath: testFile1URL.path))
+        XCTAssertFalse(fm.contentsEqual(atPath: testFile1URL.path, andPath: testFile2URL.path))
+        XCTAssertFalse(fm.contentsEqual(atPath: testFile3URL.path, andPath: testFile4URL.path))
+        XCTAssertFalse(fm.contentsEqual(atPath: symlink, andPath: testFile1URL.path))
+        XCTAssertTrue(fm.contentsEqual(atPath: testDir1.path, andPath: testDir1.path))
+        XCTAssertTrue(fm.contentsEqual(atPath: testDir2.path, andPath: testDir3.path))
+        XCTAssertFalse(fm.contentsEqual(atPath: testDir1.path, andPath: testDir2.path))
```
Comment (kperryua) [inline TestFoundation/TestFileManager.swift]: A test case that verifies recursive (in)equality is needed.

File: Foundation/FileManager.swift [hunk 5]
```diff
@@ -637,11 +637,161 @@ open class FileManager : NSObject {
     open func isDeletableFile(atPath path: String) -> Bool {
         NSUnimplemented()
     }
-    
+
+    private func _compareFiles(withFileSystemRepresentation file1Rep: UnsafePointer<Int8>, andFileSystemRepresentation file2Rep: UnsafePointer<Int8>, size: Int64) -> Bool {
+        let bufSize = min(size, 1024 * 1024)
```
Comment (kperryua) [inline Foundation/FileManager.swift]: It might be worth using `st_blksize` to pick an optimal block size. 1 MB (*2) seems like a rather large allocation.

Comment (spevans) [reply Foundation/FileManager.swift]: Im not sure how to relate it to block size, given there are only a few fixed block sizes. How does Darwin's Foundation decide on the buffer size to use?

**Comment by spevans** [comment]
@swift-ci please test

File: Foundation/FileManager.swift (hunk 5 above)
Comment (kperryua) [reply Foundation/FileManager.swift]: `st_blksize` is documented as "The optimal I/O block size for the file". I've taken that to mean that, especially if we're not reading the entire contents of the file into a single buffer (a la NSData), that this value is the system-declared preferred buffer size for I/O operations. I won't link directly to any here, but various implementations of 'cat' agree (including Darwin's).

Darwin's implementation of this method is fairly old and happens not to use this value—opting for a constant 8K size instead—though it should probably should.

Comment (spevans) [reply Foundation/FileManager.swift]: Ah right, I mis-read this as using some multiple of `st_blksize` rather than the value directly.

**Comment by spevans** [comment]
@swift-ci please test

**Comment by alblue** [comment]
@swift-ci please test

**Comment by spevans** [comment]
I added some extra tests for deep subdirectories to check the recursion. It exposed a bug in `copyItem(atPath:toPath:)` (#1515) which will need to be merged before this and also I came across a bug in `fileExists(atPath:isDirectory)` (#1514)  which should also go in .

**Comment by parkera** [comment]
Great, thanks @spevans!

**Comment by millenomi** [comment]
@swift-ci please test

**Pull request merged**